### ISSUE TITLE
Need help understanding Go

### ISSUE THREAD

**Issue by tylerflint** [nanobox-io/nanobox-engine-golang#5]
[Nanobox Desktop](https://desktop.nanobox.io) automates the creation of isolated, virtual development environments. It scans your code and builds a suitable environment within a VM. In order to best automate the build of Go environments, we would greatly appreciate it if you would be willing to answer some or all of the following questions, in as much detail as you can. There are no wrong answers:
### Detection
- What is the most accurate way you can think of to detect a Go app?
### Environment
- Which Go runtimes need to be available to your app?
- Are any other runtimes (nodejs, python, etc) needed?
- Are any utilities, webservers, or other binaries (apache, nginx, etc) needed?
- How do you manage/install dependencies?
- Where are your dependencies stored? (which directories)
### Service detection
- Does your app depend on services like postgres, redis, etc?
- How could we detect that your app depends on them?
- Does your app require any special configuration for these services?
### Application configuration
- Does your app require special configuration via config files or environment variables?
- Where and how could we configure your app to connect to the services that were previously detected?
### Compilation/Build
- Does your app need to be compiled?
- Is there a build process required to generate a release?
- Ultimately, what should be copied into the live environment? (all source code, specific directories, etc)
### Finally, running your app
- Is your app strictly a web (http) app, or does it provide a service on tcp/udp as well?
- How do you run your app?

Thanks in advance for sharing your time and expertise!

**Comment by Gufran** [comment]
A lot of these answers are opinionated and different people prefer different tools to configure the development environment so I'm going to try and generalise what I can.

##### What is the most accurate way you can think of to detect a Go app?

Scanning for `.go` files is sufficient.

##### Which Go runtimes need to be available to your app?

One should be able to define the version number in boxfile and have the appropriate runtime available

##### Are any other runtimes (nodejs, python, etc) needed?

Dependency on other language in Go is not a usual case. I can think of nodejs in some cases but then there wouldn't be an automated build. I think it is safe to assume that dependency on other language introduces one more step in build process and requires custom logic (such as a Makefile) to produce the build, in which case one should be able to override the build command in boxfile.

##### Are any utilities, webservers, or other binaries (apache, nginx, etc) needed?

Nope

##### How do you manage/install dependencies?

Some people use tools like [govendor](https://github.com/kardianos/govendor), some prefer to manually copy the libraries in vendor directory. Govendor, among other tools, is capable of managing dependencies in vendor directories and one can enable or disable the vendor experiment by setting `GO15VENDOREXPERIMENT` to 1 or 0 (in boxfile, maybe?)

##### Where are your dependencies stored? (which directories)

Depending on the tool used it can be internals, godeps, deps, and a number of other names, but if vendor experiment is enabled (with `GO15VENDOREXPERIMENT`) it will always be `vendor`

##### Does your app depend on services like postgres, redis, etc?

Most apps will depend on some kind of data storage system, it should be configurable in boxfile

##### How could we detect that your app depends on them?

By checking `.go` files for package imports. An application that has dependencies vendored in can be scanned and if any file including those in the vendor directory imports the package `database/sql` then the app depends on a SQL storage, to detect other storages such as redis and NoSQL solutions it will require knowledge of those other libraries.

##### Does your app require any special configuration for these services?

Nope

##### Does your app require special configuration via config files or environment variables?

If it does, configuration can either be stored in a file or on remote provider such as consul or vault. File path or network address can be configured in boxfile and exposed in environment variables

##### Where and how could we configure your app to connect to the services that were previously detected?

Configuration file, consul, environment variables and command line parameters are the only configuration points I can think of.

##### Does your app need to be compiled?

Yes

##### Is there a build process required to generate a release?

In most cases it is usually `go build` command but in some cases it can be a make task such as `make build`, it should be configurable in boxfile.

##### Ultimately, what should be copied into the live environment? (all source code, specific directories, etc)

The binary produced as a result of compilation

##### Is your app strictly a web (http) app, or does it provide a service on tcp/udp as well?

It can be a web app but that is not always true, or it can expose metrics over tcp but that is not always true either. 

##### How do you run your app?

Execute the binary with command line flags

**Comment by sdomino** [comment]
I think @Gufran pretty much covered it. I'll just add my personal setup (opinion).

From my `/User/<username>/` directory I created a `/gocode` folder and then pretty much followed [this guide](https://golang.org/doc/code.html) on setting up the rest of my go environment and path.

Also, something I do (for the time being) is I created a `/godeps` folder at the same location as `/gocode` and put it first in my path, so that when ever I do a `go get` it will load any 3rd party libraries in there, rather than mixed with my projects.
